run_test.py

```diff
@@ -72,78 +87,108 @@ def build_fn(atom_features: int = 64, message_steps: int = 8,
 
-    valid_data = pd.read_csv('data/valid.csv')
+    valid_data = pd.read_csv(data_dir / 'valid.csv')
     valid_loader = make_data_loader(valid_data['smiles'], valid_data['output'], batch_size=args.batch_size,
                                     max_size=args.padded_size, drop_last_batch=True)
-    ic50s = np.concatenate([x[1].numpy() for x in iter(train_loader)], axis=0)
+
+    # Determine the amount of scaling to provide
+    y_train = train_data['output']
+    y_scale = y_train / train_data['n_atom']  # Output layer scales atomic contributions for atomic-contrib networks
+
+    y_scale_mean = y_scale.mean()
+    y_scale_std = y_scale.std()
+
     with strategy.scope():
-    # Make the model
-        model = build_fn(atom_features=args.atom_features, message_steps=args.num_messages,output_layers=args.output_layers)
+        # Make the model
+        model = build_fn(atom_features=args.atom_features, message_steps=args.num_messages,
+                         output_layers=args.output_layers)
 
         # Set the scale for the output parameter
-        #ic50s = np.concatenate([x[1].numpy() for x in iter(train_loader)], axis=0)
-        model.get_layer('scale').set_weights([np.array([[ic50s.std()]]), np.array([ic50s.mean()])])
+        model.get_layer('scale').set_weights([np.array([[y_scale_mean]]), np.array([y_scale_std])])
 
-        # Train the model
-        final_learn_rate = 1e-6
-        init_learn_rate = 1e-3
+        # Configure the LR schedule
+        init_learn_rate = args.lr_start
+        final_learn_rate = init_learn_rate * 1e-3
         decay_rate = (final_learn_rate / init_learn_rate) ** (1. / (args.num_epochs - 1))
 
+
         def lr_schedule(epoch, lr):
             return lr * decay_rate
 
 
+        # Compile the model then train
         model.compile(Adam(init_learn_rate), 'mean_squared_error', metrics=['mean_absolute_error'])
+        start_time = perf_counter()
         history = model.fit(
             train_loader, validation_data=valid_loader, epochs=args.num_epochs, verbose=True,
-            shuffle=False, callbacks=[
-                LRLogger(),
-                EpochTimeLogger(),
+            shuffle=False,
+            callbacks=[
                 cb.LearningRateScheduler(lr_schedule),
-                cb.ModelCheckpoint(os.path.join(test_dir, 'best_model.h5'), save_best_only=True),
-                cb.EarlyStopping(patience=128, restore_best_weights=True),
-                cb.CSVLogger(os.path.join(test_dir, 'train_log.csv')),
+                cb.ModelCheckpoint(test_dir / 'best_model.h5', save_best_only=True),
+                # We restart the best weights, but do not halt early to simplify timing across
+                cb.EarlyStopping(patience=args.num_epochs, restore_best_weights=True),
+                cb.CSVLogger(test_dir / 'train_log.csv'),
                 cb.TerminateOnNaN()
-            ], steps_per_epoch=10
+            ],
```

Comment:
steps_per_epoch is required for running on IPU. Otherwise all other changes looks ok. I am merging them to another branch and if it works fine on IPU- then we intend to merge the changes to the master.

Reply:
Sounds like a great plan. Sorry, I should have asked about the "steps_per_epoch" part as it seemed intentional?

Would you like me to detect the number of steps rather than keeping it fixed at a small number?

Reply:
yes, predetermining is the right way to do. Just to test that out initially, we use a fixed number.

Reply:
Makes sense :) I'll get you that change soon